File: 664. Strange Printer/Solution.java
class Solution {
    public int strangePrinter(String s) {
        final int n = s.length();
        int[][] dp = new int[n][n];
        return strangePrinter(s, 0, n - 1, dp);
    }
    private int strangePrinter(final String s, int i, int j, int[][] dp) {
        if (i > j)
            return 0;
        if (dp[i][j] > 0)
            return dp[i][j];
        dp[i][j] = strangePrinter(s, i + 1, j, dp) + 1;
        for (int k = i + 1; k <= j; ++k) {
            if (s.charAt(k) == s.charAt(i)) {
                dp[i][j] = Math.min(dp[i][j], strangePrinter(s, i, k - 1, dp) + strangePrinter(s, k + 1, j, dp));
            }
        }
        return dp[i][j];
    }
}

File: 664. Strange Printer/Solution.py
class Solution:
    def strangePrinter(self, s: str) -> int:
        if not s:
            return 0
        n = len(s)
        dp = [[n for _ in range(n)] for _ in range(n)]
        for i in range(n):
            dp[i][i] = 1
        for j in range(n):
            for i in range(j, -1, -1):
                for k in range(i, j):
                    dp[i][j] = min(dp[i][j], dp[i][k] + dp[k + 1][j] - (1 if s[k] == s[j] else 0))
        return dp[0][n - 1]